File: 003-Longest-Substring-Without-Repeating-Characters/Solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    var _map = {};
    var left = 0;
    var _max = 0;
    for(var i = 0, l = s.length; i < l; ++i) {
        if(_map[s[i]] !== undefined) {
            left = left > _map[s[i]] + 1 ? left : _map[s[i]] + 1;
        }
        _max = _max > i - left + 1 ? _max : i - left + 1;
        _map[s[i]] = i;
    }
    
    return _max;
};

File: 003-Longest-Substring-Without-Repeating-Characters/Solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, unsigned> _map;
        int a, _max;
        a = 0;
        _max = 0;
        
        for(int i = 0; i < s.length(); ++i) {
            char c = s[i];
            if(_map.find(c) != _map.end()) {
                a = max(int(_map[c] + 1), a);
            }
            _max = max(_max, i - a + 1);
            _map[c] = i;
        }
        
        return _max;
    }
};
